File: spring/src/main/java/org/sw/sample/GlobalExceptionHandler.java
package org.sw.sample;

import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.sw.sample.exception.InternalServerError;
import org.sw.sample.exception.NotFoundException;

import java.net.URI;

@RestControllerAdvice
public class GlobalExceptionHandler {

    private final Logger logger = LoggerFactory.getLogger(this.getClass());


    @ExceptionHandler(NotFoundException.class)
    public ResponseEntity<ProblemDetail> handlerNotFoundExceptions(HttpServletRequest request,
                                                                   NotFoundException cause) {
        try {
            ProblemDetail problem = ProblemDetail.forStatus(cause.getStatus().value());
            problem.setType(URI.create(request.getRequestURI()));
            problem.setTitle(cause.getReason());
            problem.setDetail(cause.getMessage());
            problem.setInstance(null);  // api-documents
            return ResponseEntity
                    .status(problem.getStatus())
                    .header("Content-Type", "application/problem+json")
                    .body(problem);
        } finally {
            print(cause);
        }
    }


    @ExceptionHandler(InternalServerError.class)
    public ResponseEntity<ProblemDetail> handleInternalServerErrors(HttpServletRequest servlet, InternalServerError cause) {
        try {
            ProblemDetail problem = ProblemDetail.forStatus(500);
            problem.setType(URI.create(servlet.getRequestURI())); // Request Path
            problem.setTitle(cause.getReason());
            problem.setDetail(cause.getMessage());
            problem.setInstance(null);  // api-documents
            problem.setProperties(null); // 3rd-sys messages
            return ResponseEntity
                    .status(problem.getStatus())
                    .header("Content-Type", "application/problem+json")
                    .body(problem);
        } finally {
            print(cause);
        }
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ProblemDetail> handleGlobalExceptions(HttpServletRequest servlet, Exception cause) {
        try {
            ProblemDetail problem = ProblemDetail.forStatus(500);
            problem.setType(URI.create(servlet.getRequestURI())); // Request Path
            problem.setTitle(HttpStatus.INTERNAL_SERVER_ERROR.getReasonPhrase());

            return ResponseEntity
                    .status(problem.getStatus())
                    .header("Content-Type", "application/problem+json")
                    .body(problem);
        } finally {
            print(cause);
        }
    }

    private void print(Exception exception) {
        logger.info("Exception Occurred: {} - {}", exception.getClass().getSimpleName(), exception.getMessage());
        logger.error("Exception Occurred", exception);
    }

}

File: kotlin/src/main/kotlin/org/sw/spring/common/GlobalExceptionHandler.kt
package org.sw.spring.common

import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.sw.spring.common.exception.CommonException
import java.net.URI

@RestControllerAdvice
class GlobalExceptionHandler {

    companion object {
        private val logger = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)
    }

    @ExceptionHandler(CommonException::class)
    fun handleCommonExceptions(
        servlet: HttpServletRequest,
        cause: CommonException
    ): ResponseEntity<ProblemDetail> {
        return ProblemDetail.forStatus(cause.status).apply {
            type = URI.create(servlet.requestURI)
            title = cause.title
            detail = cause.detail
            instance = URI.create(cause.instance)
        }.let {
            ResponseEntity.status(it.status)
                .header("Content-Type", "application/problem+json")
                .body(it)
        }
    }

    @ExceptionHandler(Exception::class)
    fun handleExceptions(
        servlet: HttpServletRequest,
        cause: Exception
    ): ResponseEntity<ProblemDetail> {
        return ProblemDetail.forStatus(500).apply {
            type = URI.create(servlet.requestURI)
            title = HttpStatus.INTERNAL_SERVER_ERROR.reasonPhrase
            detail = HttpStatus.INTERNAL_SERVER_ERROR.reasonPhrase
            instance = null
        }.let {
            ResponseEntity.status(it.status)
                .header("Content-Type", "application/problem+json")
                .body(it)
        }
    }




}
